src/mast/mod.rs

```diff
@@ -987,11 +1045,8 @@ impl<B: Backend> Mast<B> {
         // evaluate generic return types using inferred values
         let ret_ty = match &fn_sig.return_type {
```

Comment:
this will need to be recursive also (as I pointed in the other PR), but unrelated to this PR